pkg/incentives/incentives.go

```diff
@@ -0,0 +1,371 @@
+	backend  ChainBackend
+	monitor  Monitor
+	contract IncentivesContract
+	reserve  postage.Storer
+	sampler  Sampler
+	overlay  swarm.Address
+	slashedC chan struct{}
+	quit     chan struct{}
+	wg       sync.WaitGroup
+}
+
+func New(
+	overlay swarm.Address,
+	backend ChainBackend,
+	logger log.Logger,
+	monitor Monitor,
+	incentives IncentivesContract,
+	reserve postage.Storer,
+	sampler Sampler,
+	blockTime time.Duration, blockPerRound, blocksPerPhase uint64) *Agent {
+
+	s := &Agent{
+		overlay:  overlay,
+		backend:  backend,
+		logger:   logger.WithName(loggerName).Register(),
+		contract: incentives,
+		reserve:  reserve,
+		monitor:  monitor,
+		sampler:  sampler,
+		slashedC: make(chan struct{}),
+		quit:     make(chan struct{}),
+	}
+
+	s.wg.Add(1)
+	go s.start(blockTime, blockPerRound, blocksPerPhase)
+
+	return s
+}
+
+// start polls the current block number, calculates, and publishes only once the current phase.
+// Each round is blocksPerRound long and is divided in to three blocksPerPhase long phases: commit, reveal, claim.
+// The sample phase is triggered upon entering the claim phase and may run until the end of the commit phase.
+// If our neighborhood is selected to participate, a sample is created during the sample phase. In the commit phase,
+// the sample is submitted, and in the reveal phase, the obfuscation key from the commit phase is submitted.
+// Next, in the claim phase, we check if we've won, and the cycle repeats. The cycle must occur in the length of one round.
+func (s *Agent) start(blockTime time.Duration, blocksPerRound, blocksPerPhase uint64) {
+
+	defer s.wg.Done()
+
+	var (
+		mtx            sync.Mutex
+		sampleRound    uint64 = math.MaxUint64
+		commitRound    uint64 = math.MaxUint64
+		revealRound    uint64 = math.MaxUint64
+		round          uint64
+		reserveSample  []byte
+		obfuscationKey []byte
+		storageRadius  uint8
+		phaseEvents    = newEvents()
+		sampleEvents   = newEvents()
+	)
+
+	// cancel all possible running phases
+	defer phaseEvents.Close()
+	defer sampleEvents.Close()
+
+	commitF := func(ctx context.Context) {
+		phaseEvents.Cancel(claim)
+
+		mtx.Lock()
+		round := round
+		sampleRound := sampleRound
+		storageRadius := storageRadius
+		reserveSample := reserveSample
+		mtx.Unlock()
+
+		if round-1 == sampleRound { // the sample has to come from previous round to be able to commit it
+			obf, err := s.commit(ctx, storageRadius, reserveSample)
+			if err != nil {
+				s.logger.Error(err, "commit")
+			} else {
+				mtx.Lock()
+				obfuscationKey = obf
+				commitRound = round
+				mtx.Unlock()
+				s.logger.Debug("commit phase")
+			}
+		}
+	}
+
+	// when the sample finishes, if we are in the commit phase, run commit
+	sampleEvents.On(sampleEnd, func(ctx context.Context) {
+		if phaseEvents.Last() == commit {
+			commitF(ctx)
+		}
+	})
+
+	// when we enter the commit phase, if the sample is already finished, run commit
+	phaseEvents.On(commit, func(ctx context.Context) {
```

Comment:
It could happen that both this and the above event would be triggered and we have 2 go routines running the commitF. If we have these separate events, I am not sure if this can be done cleanly.